File: mocida/src/uikit/theme.c
#include <uikit/theme.h>

static UITheme g_theme;
static int     g_themeInited = 0;

static void EnsureInited(void) {
    if (g_themeInited) return;
    UITheme_FillLight(&g_theme);
    g_themeInited = 1;
}

void UITheme_FillLight(UITheme* out) {
    if (!out) return;
    out->primary    = (UIColor){ 59, 130, 246, 1.0f };
    out->onPrimary  = (UIColor){ 255, 255, 255, 1.0f };
    out->surface    = (UIColor){ 255, 255, 255, 1.0f };
    out->onSurface  = (UIColor){ 15, 23, 42, 1.0f };
    out->background = (UIColor){ 248, 250, 252, 1.0f };
    out->border     = (UIColor){ 203, 213, 225, 1.0f };
    out->success    = (UIColor){ 34, 197, 94, 1.0f };
    out->warning    = (UIColor){ 250, 204, 21, 1.0f };
    out->danger     = (UIColor){ 239, 68, 68, 1.0f };
    out->disabled   = (UIColor){ 148, 163, 184, 1.0f };
    out->radius     = 8.0f;
    out->spacing    = 8.0f;
    out->fontSizeSmall  = 12.0f;
    out->fontSizeMedium = 16.0f;
    out->fontSizeLarge  = 24.0f;
}

void UITheme_FillDark(UITheme* out) {
    if (!out) return;
    out->primary    = (UIColor){ 96, 165, 250, 1.0f };
    out->onPrimary  = (UIColor){ 15, 23, 42, 1.0f };
    out->surface    = (UIColor){ 30, 41, 59, 1.0f };
    out->onSurface  = (UIColor){ 226, 232, 240, 1.0f };
    out->background = (UIColor){ 15, 23, 42, 1.0f };
    out->border     = (UIColor){ 51, 65, 85, 1.0f };
    out->success    = (UIColor){ 74, 222, 128, 1.0f };
    out->warning    = (UIColor){ 253, 224, 71, 1.0f };
    out->danger     = (UIColor){ 248, 113, 113, 1.0f };
    out->disabled   = (UIColor){ 100, 116, 139, 1.0f };
    out->radius     = 8.0f;
    out->spacing    = 8.0f;
    out->fontSizeSmall  = 12.0f;
    out->fontSizeMedium = 16.0f;
    out->fontSizeLarge  = 24.0f;
}

const UITheme* UITheme_GetGlobal(void) {
    EnsureInited();
    return &g_theme;
}

void UITheme_SetGlobal(const UITheme* t) {
    if (!t) return;
    g_theme = *t;
    g_themeInited = 1;
}

File: mocida-rs/mocida/src/theme.rs
//! `UITheme` — design tokens consumed by widgets on construction.

use mocida_sys as sys;

use crate::color::Color;

/// Snapshot of every design token mocida exposes.
#[derive(Debug, Clone, Copy)]
pub struct Theme {
    /// Primary brand color (buttons, sliders, focused borders).
    pub primary: Color,
    /// Text on primary surfaces.
    pub on_primary: Color,
    /// Card / sheet / dialog background.
    pub surface: Color,
    /// Text on surface.
    pub on_surface: Color,
    /// Window background.
    pub background: Color,
    /// Borders and dividers.
    pub border: Color,
    /// Success tint.
    pub success: Color,
    /// Warning tint.
    pub warning: Color,
    /// Destructive / error tint.
    pub danger: Color,
    /// Disabled / inactive tint.
    pub disabled: Color,
    /// Default rounded-corner radius.
    pub radius: f32,
    /// Default gap between things in a stack.
    pub spacing: f32,
    /// Small font size token.
    pub font_size_small: f32,
    /// Default font size token.
    pub font_size_medium: f32,
    /// Large font size token.
    pub font_size_large: f32,
}

impl Theme {
    /// Returns mocida's default light theme.
    pub fn light() -> Self {
        let mut raw = sys::UITheme::default();
        unsafe { sys::UITheme_FillLight(&mut raw) };
        Self::from_raw(raw)
    }

    /// Returns mocida's default dark theme.
    pub fn dark() -> Self {
        let mut raw = sys::UITheme::default();
        unsafe { sys::UITheme_FillDark(&mut raw) };
        Self::from_raw(raw)
    }

    /// Reads the current global theme.
    pub fn global() -> Self {
        // SAFETY: mocida guarantees the pointer is non-null and lives
        // for the process lifetime.
        let raw = unsafe { *sys::UITheme_GetGlobal() };
        Self::from_raw(raw)
    }

    /// Installs this theme as the new global. Existing widgets keep
    /// their current colors; new ones pick up the new tokens.
    pub fn install(self) {
        let raw = self.into_raw();
        unsafe { sys::UITheme_SetGlobal(&raw) };
    }

    fn from_raw(raw: sys::UITheme) -> Self {
        Self {
            primary: Color::from_raw(raw.primary),
            on_primary: Color::from_raw(raw.onPrimary),
            surface: Color::from_raw(raw.surface),
            on_surface: Color::from_raw(raw.onSurface),
            background: Color::from_raw(raw.background),
            border: Color::from_raw(raw.border),
            success: Color::from_raw(raw.success),
            warning: Color::from_raw(raw.warning),
            danger: Color::from_raw(raw.danger),
            disabled: Color::from_raw(raw.disabled),
            radius: raw.radius,
            spacing: raw.spacing,
            font_size_small: raw.fontSizeSmall,
            font_size_medium: raw.fontSizeMedium,
            font_size_large: raw.fontSizeLarge,
        }
    }

    fn into_raw(self) -> sys::UITheme {
        sys::UITheme {
            primary: self.primary.into_raw(),
            onPrimary: self.on_primary.into_raw(),
            surface: self.surface.into_raw(),
            onSurface: self.on_surface.into_raw(),
            background: self.background.into_raw(),
            border: self.border.into_raw(),
            success: self.success.into_raw(),
            warning: self.warning.into_raw(),
            danger: self.danger.into_raw(),
            disabled: self.disabled.into_raw(),
            radius: self.radius,
            spacing: self.spacing,
            fontSizeSmall: self.font_size_small,
            fontSizeMedium: self.font_size_medium,
            fontSizeLarge: self.font_size_large,
        }
    }
}
